### PR TITLE
Fixing and Expanding Label Functionality

### PR DESCRIPTION
**Pull request by bkendall**
A few things in this PR

- removed the `misc` folder in favor of `resources`
- consolidated the `models` and `utils` tests into a `unit` folder (makes the testing commands more clear)
- always store labels as Objects (single keys with no value are represented with a value of empty string)
- allow filtering of containers by labels
- allow filtering of containers by status (but only by `running` or not - other statuses not supported atm)

Tests have been added/expanded.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -40,12 +42,24 @@ app.use(function (req, res, next) {
...
+  function (req, res, next) {
+    // this can possibly throw, so wrapping in Promise.
+    Promise.try(function () {
+      if (req.query && req.query.filters) {
+        // need to JSON.parse the filters, first.
+        req.query.filters = JSON.parse(req.query.filters)
+        req.query.filters = ContainerStore._formatQueryFilters(req.query.filters)
+      }
+    })
+      .asCallback(next)
```
Comment (anandkumarpatel) [inline lib/index.js]: nit: spacing? or is this on purpose

File: lib/models/container-store.js [hunk 2]
```diff
@@ -57,7 +60,28 @@ ContainerStore.prototype.deleteById = function (id) {
     })
 }
 
-ContainerStore.prototype.listContainers = function () {
+ContainerStore._formatQueryFilters = function (filters) {
+  if (filters.label) {
```
Comment (anandkumarpatel) [inline lib/models/container-store.js]: `101/isArray` ?

Comment (anandkumarpatel) [inline lib/models/container-store.js]: nm, not that kind of array

File: lib/models/container-store.js [hunk 3]
```diff
@@ -75,10 +99,57 @@ ContainerStore.prototype.listContainers = function () {
...
+    switch (filter) {
+      case 'label':
+        var key = Object.keys(filters[filter])[0]
+        var value = filters[filter][key]
+        containers = containers.filter(filterLabel(key, value))
+        break
+      case 'status':
+        var running = filters[filter] === 'running'
+        containers = containers.filter(filterStatus(running))
+        break
```
Comment (anandkumarpatel) [inline lib/models/container-store.js]: add a throw to default case? then you can return the user `400`

File: lib/models/container-store.js [hunk 4]
```diff
@@ -75,10 +99,57 @@ ContainerStore.prototype.listContainers = function () {
...
+    return function (c) {
+      if (!c.Labels) { return false }
+      var containerValue = c.Labels[key]
+      if (containerValue === '' && value === '') { return true }
+      if (!containerValue || containerValue !== value) { return false }
+      return true
+    }
+  }
+  function filterStatus (status) {
+    return function (c) { return c.State.Running === status }
```
Comment (anandkumarpatel) [inline lib/models/container-store.js]: docker State json
```
    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "2015-12-11T19:12:31.191598747Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 30821,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2015-12-11T19:17:31.460165195Z"
    },
```
state is based on key?

File: test/functional/containers.js [hunk 5]
```diff
@@ -85,35 +85,85 @@ describe('containers', function () {
...
+      docker.listContainers(opts, function (err, containers) {
+        if (err) { return done(err) }
+        assert.lengthOf(containers, 1)
+        // dockerode uses .id, not .Id
+        assert.equal(containers[0].Id, container.id)
+        done()
+      })
+    })
+
+    it('should filter on Labels with complex values', function (done) {
```
Comment (anandkumarpatel) [inline test/functional/containers.js]: add a test for misspelled `label`?

**Comment by anandkumarpatel** [comment]
solid PR few minors

File: lib/index.js (hunk 1 above)
Comment (bkendall) [inline lib/index.js]: on purpose (tabbed in the `.` after `Promise.try()`

File: lib/models/container-store.js (hunk 3 above)
Comment (bkendall) [inline lib/models/container-store.js]: passing an invalid query parameter isn't a `400` error, it's just not handled (at least, that's my understanding).

File: lib/models/container-store.js (hunk 4 above)
Comment (bkendall) [inline lib/models/container-store.js]: I'm not sure what you're getting at. I do mention in the description above that `running` is the only supported filter at the moment. I'm sure I can write out the rest of the states eventually, but I don't see a need for it at the very moment

File: test/functional/containers.js (hunk 5 above)
Comment (bkendall) [inline test/functional/containers.js]: again, just wouldn't be handled, I imagine. you can pass garbage in query strings all the time - it's not an error

File: lib/models/container-store.js (hunk 2 above)
Comment (bkendall) [inline lib/models/container-store.js]: `isArray` isn't a thing, but I added a check

**Pull request merged**